File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/longestZigZag_1372.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.TreeNode;

public class longestZigZag_1372 {
    int res = 0;

    public int longestZigZag(TreeNode root) {
        helper(root, false, false, 0);
        return res;
    }

    private void helper(TreeNode root, boolean left, boolean right, int path) {
        if (root == null) return;
        res = Math.max(res, path);
        helper(root.left, true, false, right ? path + 1 : 1);
        helper(root.right, false, true, left ? path + 1 : 1);
    }
}

File: Clion/ZeroTrac/Tree/longestZigZag_1372.cpp
//
// Created by TzashiNorpu on 10/27/2022.
//
#include "tree.h"

using namespace Tree;

int Solution::longestZigZag(TreeNode *root) {
  int res = 0;
  function<void(TreeNode *, bool, bool, int)> helper = [&](TreeNode *node, bool
  left, bool right, int path) -> void {
      if (!node) return;
      res = max(res, path);
      helper(node->left, true, false, right ? path + 1 : 1);
      helper(node->right, false, true, left ? path + 1 : 1);
  };
  helper(root, false, false, 0);
  return res;
}